File: crates/verify/ess-conformance/src/go/temporal.go
// Native live observation semantics. These types do not admit a new suite format.

// Occurrence retains a real observation's identity on one scenario-local monotonic clock.
type Occurrence struct {
	Sequence uint64
	AtMS     uint64
	Event    string
	Payload  map[string]Node
}

// ObservationBatch extends a subscribed lifetime with every occurrence strictly before
// CompleteBeforeMS. Sequence ordinals are contiguous across its subscribed sources.
type ObservationBatch struct {
	Lifetime         string
	After            uint64
	CompleteBeforeMS uint64
	Occurrences      []Occurrence
}

// OccurrenceMatcher compares an event name and each named actual payload field.
type OccurrenceMatcher struct {
	Event   string
	Payload map[string]Node
}

func (m OccurrenceMatcher) matches(item Occurrence) bool {
	if m.Event != item.Event {
		return false
	}
	for field, value := range m.Payload {
		actual, exists := lookup(item.Payload, field)
		if !exists || !responseEqual(actual, value) {
			return false
		}
	}
	return true
}

func temporalFields(fields map[string]Node) (map[string]Node, error) {
	raw, err := json.Marshal(fields)
	if err != nil {
		return nil, &TemporalError{Kind: "invalid_claim", Reason: "payload is not a typed wire value"}
	}
	var copied map[string]Node
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	if err := decoder.Decode(&copied); err != nil {
		return nil, &TemporalError{Kind: "invalid_claim", Reason: "payload cannot be retained"}
	}
	return copied, nil
}

func (m OccurrenceMatcher) checked() (OccurrenceMatcher, error) {
	if !qualifiedName.MatchString(m.Event) {
		return m, &TemporalError{Kind: "invalid_claim", Reason: "invalid event name"}
	}
	fields, err := temporalFields(m.Payload)
	return OccurrenceMatcher{Event: m.Event, Payload: fields}, err
}

// TemporalError preserves incomplete observation separately from a real counterexample.
type TemporalError struct {
	Kind           string
	Reason         string
	Anchor         uint64
	RequiredMS     uint64
	ObservedMS     uint64
	Before         uint64
	After          uint64
	Counterexample *Occurrence
}

func (e *TemporalError) Error() string {
	return fmt.Sprintf("temporal %s: %s (anchor=%d, required=%d, observed=%d)", e.Kind, e.Reason, e.Anchor, e.RequiredMS, e.ObservedMS)
}

// ObservationLedger is native-owned history. A source failure permanently invalidates it.
// It is owned by one evaluator goroutine; adapters synchronize their own incoming sources.
type ObservationLedger struct {
	lifetime         string
	occurrences      []Occurrence
	completeBeforeMS uint64
	broken           string
}

// NewObservationLedger starts observation before any stimulus is issued.
func NewObservationLedger(lifetime string) (*ObservationLedger, error) {
	if strings.TrimSpace(lifetime) == "" {
		return nil, &TemporalError{Kind: "incomplete", Reason: "empty observation lifetime"}
	}
	return &ObservationLedger{lifetime: lifetime}, nil
}

// Invalidate records the first gap, including disconnect, overflow, decode or cancellation.
func (l *ObservationLedger) Invalidate(reason string) error {
	if l.broken == "" {
		if reason == "" {
			reason = "unspecified observation failure"
		}
		l.broken = reason
	}
	return l.healthy()
}

func (l *ObservationLedger) healthy() error {
	if l.broken != "" {
		return &TemporalError{Kind: "incomplete", Reason: l.broken}
	}
	return nil
}

// Cursor is the last admitted ordinal. It cannot advance on a partly valid batch.
func (l *ObservationLedger) Cursor() uint64 {
	if len(l.occurrences) == 0 {
		return 0
	}
	return l.occurrences[len(l.occurrences)-1].Sequence
}

// Accept snapshots and admits a complete batch atomically. Malformed batches poison the
// lifetime; a later good frame cannot conceal an earlier gap.
func (l *ObservationLedger) Accept(batch ObservationBatch) error {
	if err := l.healthy(); err != nil {
		return err
	}
	raw, err := json.Marshal(batch)
	if err != nil {
		return l.Invalidate("observation is not a typed wire value")
	}
	var copied ObservationBatch
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	if err := decoder.Decode(&copied); err != nil {
		return l.Invalidate("observation cannot be retained")
	}
	if copied.Lifetime != l.lifetime {
		return l.Invalidate("observation lifetime changed")
	}
	if copied.After != l.Cursor() {
		return l.Invalidate("observation cursor is not the preceding receipt")
	}
	if copied.CompleteBeforeMS < l.completeBeforeMS {
		return l.Invalidate("observation watermark regressed")
	}
	previous, at := l.Cursor(), l.completeBeforeMS
	for _, item := range copied.Occurrences {
		if !qualifiedName.MatchString(item.Event) {
			return l.Invalidate("invalid event name")
		}
		if previous == ^uint64(0) || item.Sequence != previous+1 {
			return l.Invalidate("missing or duplicate occurrence ordinal")
		}
		if item.AtMS < at || item.AtMS >= copied.CompleteBeforeMS {
			return l.Invalidate("occurrence falls outside the complete monotonic interval")
		}
		previous, at = item.Sequence, item.AtMS
	}
	l.occurrences = append(l.occurrences, copied.Occurrences...)
	l.completeBeforeMS = copied.CompleteBeforeMS
	return nil
}

func (l *ObservationLedger) anchor(sequence uint64) (*Occurrence, error) {
	if err := l.healthy(); err != nil {
		return nil, err
	}
	for index := range l.occurrences {
		if l.occurrences[index].Sequence == sequence {
			return &l.occurrences[index], nil
		}
	}
	return nil, &TemporalError{Kind: "unknown_anchor", Anchor: sequence}
}

func copyOccurrence(item Occurrence) *Occurrence {
	// The source is an already admitted wire snapshot. Copy again so diagnostics and public
	// lookup results cannot mutate the evaluator's retained history.
	raw, err := json.Marshal(item)
	if err != nil {
		panic("native occurrence snapshot lost wire shape")
	}
	var copied Occurrence
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	if err := decoder.Decode(&copied); err != nil {
		panic("native occurrence snapshot cannot be copied")
	}
	return &copied
}

// Find returns the first actual matching occurrence strictly after an earlier ordinal,
// or after subscription when after is zero. The caller receives an independent snapshot.
func (l *ObservationLedger) Find(matcher OccurrenceMatcher, after uint64) (*Occurrence, error) {
	if err := l.healthy(); err != nil {
		return nil, err
	}
	if after != 0 {
		if _, err := l.anchor(after); err != nil {
			return nil, err
		}
	}
	matcher, err := matcher.checked()
	if err != nil {
		return nil, err
	}
	for _, item := range l.occurrences {
		if item.Sequence > after && matcher.matches(item) {
			return copyOccurrence(item), nil
		}
	}
	return nil, nil
}

// Ordered requires distinct actual occurrences in strict order, including one clock tick.
func (l *ObservationLedger) Ordered(before, after uint64) error {
	if _, err := l.anchor(before); err != nil {
		return err
	}
	if _, err := l.anchor(after); err != nil {
		return err
	}
	if before >= after {
		return &TemporalError{Kind: "order", Before: before, After: after}
	}
	return nil
}

func (l *ObservationLedger) end(anchor, durationMS uint64) (uint64, error) {
	start, err := l.anchor(anchor)
	if err != nil {
		return 0, err
	}
	if durationMS == 0 || durationMS > ^uint64(0)-start.AtMS {
		return 0, &TemporalError{Kind: "invalid_window"}
	}
	return start.AtMS + durationMS, nil
}

func (l *ObservationLedger) complete(requiredMS uint64) error {
	if err := l.healthy(); err != nil {
		return err
	}
	if l.completeBeforeMS < requiredMS {
		return &TemporalError{Kind: "unfinished", RequiredMS: requiredMS, ObservedMS: l.completeBeforeMS}
	}
	return nil
}

// Absent checks the entire interval after the observed anchor and before its exclusive
// end. Early counterexamples fail immediately; success requires a complete watermark.
func (l *ObservationLedger) Absent(matcher OccurrenceMatcher, anchor, durationMS uint64) error {
	end, err := l.end(anchor, durationMS)
	if err != nil {
		return err
	}
	matcher, err = matcher.checked()
	if err != nil {
		return err
	}
	for _, item := range l.occurrences {
		if item.Sequence > anchor && item.AtMS < end && matcher.matches(item) {
			return &TemporalError{Kind: "counterexample", Counterexample: copyOccurrence(item)}
		}
	}
	return l.complete(end)
}

// Stable checks the observed baseline and every scoped snapshot until the exclusive end.
// Scope selects a stream such as one account/queue; required holds the claimed metric values.
func (l *ObservationLedger) Stable(scope OccurrenceMatcher, required map[string]Node, anchor, durationMS uint64) error {
	baseline, err := l.anchor(anchor)
	if err != nil {
		return err
	}
	end, err := l.end(anchor, durationMS)
	if err != nil {
		return err
	}
	if len(required) == 0 {
		return &TemporalError{Kind: "empty_claim"}
	}
	for field := range required {
		for scoped := range scope.Payload {
			if field == scoped || strings.HasPrefix(field, scoped+".") || strings.HasPrefix(scoped, field+".") {
				return &TemporalError{Kind: "filtered_claim"}
			}
		}
	}
	scope, err = scope.checked()
	if err != nil {
		return err
	}
	required, err = temporalFields(required)
	if err != nil {
		return err
	}
	if !scope.matches(*baseline) {
		return &TemporalError{Kind: "counterexample", Counterexample: copyOccurrence(*baseline)}
	}
	for _, item := range l.occurrences {
		if item.Sequence < anchor || item.AtMS >= end || !scope.matches(item) {
			continue
		}
		for field, value := range required {
			actual, exists := lookup(item.Payload, field)
			if !exists || !responseEqual(actual, value) {
				return &TemporalError{Kind: "counterexample", Counterexample: copyOccurrence(item)}
			}
		}
	}
	return l.complete(end)
}

File: crates/verify/ess-conformance/src/temporal.rs
//! Occurrence identity and complete observation intervals for live conformance.
//!
//! Targets report append-only facts on one scenario-local monotonic clock. They do not
//! report whether an assertion passed. A watermark is exclusive: every occurrence before
//! it has been delivered. Disconnects, overflow and decoding failures invalidate the
//! lifetime; reconnecting cannot repair its history. These primitives do not extend the
//! persisted suite vocabulary by themselves.

use crate::scenario::EventRef;
use ess_primitives::node::Node;
use std::collections::BTreeMap;

/// One observed occurrence. Equal event names and payloads still have distinct identities.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Occurrence {
    /// Contiguous ordinal across all sources in this observation lifetime, starting at one.
    pub sequence: u64,
    /// Milliseconds on the shared scenario-local monotonic clock.
    pub at_ms: u64,
    /// Declared event observed through its real source.
    pub event: EventRef,
    /// Actual payload, copied at observation time.
    pub payload: BTreeMap<String, Node>,
}

/// An adapter's complete extension of one observation lifetime.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Batch {
    /// Exact lifetime returned when observation began, before stimulus.
    pub lifetime: String,
    /// Last ordinal delivered by the preceding batch, or zero initially.
    pub after: u64,
    /// Every occurrence strictly before this time is present in the ledger.
    pub complete_before_ms: u64,
    /// Newly observed facts in contiguous ordinal and monotonic time order.
    pub occurrences: Vec<Occurrence>,
}

/// A payload matcher compares only explicitly named fields, preserving absent versus null.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Matcher {
    /// Declared event name.
    pub event: EventRef,
    /// Required equal field values, resolved from literals or actual captured identities.
    pub payload: BTreeMap<String, Node>,
}

impl Matcher {
    /// Whether this occurrence has the required event and every required payload value.
    pub fn matches(&self, occurrence: &Occurrence) -> bool {
        self.event == occurrence.event
            && self
                .payload
                .iter()
                .all(|(field, value)| path_value(&occurrence.payload, field) == Some(value))
    }
}

/// Read a declared dotted field path from an actual observed payload.
pub fn path_value<'a>(payload: &'a BTreeMap<String, Node>, path: &str) -> Option<&'a Node> {
    let mut parts = path.split('.');
    let mut value = payload.get(parts.next()?)?;
    for part in parts {
        let Node::Map(fields) = value else {
            return None;
        };
        value = fields.get(part)?;
    }
    Some(value)
}

fn overlapping_paths(left: &str, right: &str) -> bool {
    left == right
        || left
            .strip_prefix(right)
            .is_some_and(|suffix| suffix.starts_with('.'))
        || right
            .strip_prefix(left)
            .is_some_and(|suffix| suffix.starts_with('.'))
}

/// A refused observation or a refuted temporal claim. Neither is a passing interval.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Error {
    /// The lifetime is not complete and cannot acquire a receipt from later good batches.
    Incomplete(String),
    /// The requested anchor does not belong to this ledger.
    UnknownAnchor(u64),
    /// A bounded assertion must claim a positive representable interval.
    InvalidWindow,
    /// A stability claim must constrain at least one payload field.
    EmptyClaim,
    /// A field being checked cannot also filter which snapshots are inspected.
    FilteredClaim,
    /// The target has not yet completed the requested window.
    Unfinished {
        /// Required exclusive watermark.
        required_ms: u64,
        /// Actual exclusive watermark.
        observed_ms: u64,
    },
    /// The real occurrence that contradicted the assertion.
    Counterexample(Occurrence),
    /// The named occurrences are not in the requested strict order.
    Order {
        /// Required earlier occurrence.
        before: u64,
        /// Required later occurrence.
        after: u64,
    },
}

/// A scenario's immutable occurrence history after native admission of each batch.
#[derive(Debug)]
pub struct Ledger {
    lifetime: String,
    occurrences: Vec<Occurrence>,
    complete_before_ms: u64,
    broken: Option<String>,
}

impl Ledger {
    /// Begin one lifetime. Reusing a ledger across reconnects or scenarios is forbidden.
    pub fn new(lifetime: String) -> Result<Self, Error> {
        if lifetime.trim().is_empty() {
            return Err(Error::Incomplete("empty observation lifetime".into()));
        }
        Ok(Self {
            lifetime,
            occurrences: Vec::new(),
            complete_before_ms: 0,
            broken: None,
        })
    }

    /// Record a source failure permanently, including loss not visible in an ordinal gap.
    pub fn invalidate(&mut self, reason: impl Into<String>) -> Error {
        let reason = self.broken.get_or_insert_with(|| reason.into());
        Error::Incomplete(reason.clone())
    }

    fn healthy(&self) -> Result<(), Error> {
        match &self.broken {
            Some(reason) => Err(Error::Incomplete(reason.clone())),
            None => Ok(()),
        }
    }

    /// Admit a complete append. A malformed batch contributes no partial facts and poisons
    /// the lifetime, so a later buffered good frame cannot conceal the gap.
    pub fn accept(&mut self, batch: Batch) -> Result<(), Error> {
        self.healthy()?;
        if let Err(reason) = self.validate(&batch) {
            return Err(self.invalidate(reason));
        }
        self.complete_before_ms = batch.complete_before_ms;
        self.occurrences.extend(batch.occurrences);
        Ok(())
    }

    fn validate(&self, batch: &Batch) -> Result<(), String> {
        if batch.lifetime != self.lifetime {
            return Err("observation lifetime changed".into());
        }
        if batch.after != self.cursor() {
            return Err("observation cursor is not the preceding receipt".into());
        }
        if batch.complete_before_ms < self.complete_before_ms {
            return Err("observation watermark regressed".into());
        }
        let mut previous = self.cursor();
        let mut at = self.complete_before_ms;
        for occurrence in &batch.occurrences {
            if previous.checked_add(1) != Some(occurrence.sequence) {
                return Err("missing or duplicate occurrence ordinal".into());
            }
            if occurrence.at_ms < at || occurrence.at_ms >= batch.complete_before_ms {
                return Err("occurrence falls outside the complete monotonic interval".into());
            }
            at = occurrence.at_ms;
            previous = occurrence.sequence;
        }
        Ok(())
    }

    /// Last accepted occurrence ordinal, for the next observation request.
    pub fn cursor(&self) -> u64 {
        self.occurrences.last().map_or(0, |item| item.sequence)
    }

    /// Current exclusive watermark, used to fence a newly activated quiet interval.
    pub fn complete_before_ms(&self) -> u64 {
        self.complete_before_ms
    }

    /// Latest scoped snapshot, after a full quiet interval beginning no earlier than
    /// the fresh activation fence. A newer scoped snapshot restarts the interval.
    pub fn quiet(
        &self,
        matcher: &Matcher,
        floor_ms: u64,
        duration_ms: u64,
    ) -> Result<&Occurrence, Error> {
        self.healthy()?;
        if duration_ms == 0 {
            return Err(Error::InvalidWindow);
        }
        let item = self
            .occurrences
            .iter()
            .rev()
            .find(|item| matcher.matches(item))
            .ok_or(Error::Unfinished {
                required_ms: 0,
                observed_ms: self.complete_before_ms,
            })?;
        let end = item
            .at_ms
            .max(floor_ms)
            .checked_add(duration_ms)
            .ok_or(Error::InvalidWindow)?;
        self.complete(end)?;
        Ok(item)
    }

    /// The first matching occurrence strictly after an earlier occurrence, or after
    /// subscription when `after` is zero. Payload matching never borrows another call's event.
    pub fn find(&self, matcher: &Matcher, after: u64) -> Result<Option<&Occurrence>, Error> {
        self.healthy()?;
        if after != 0 {
            self.anchor(after)?;
        }
        Ok(self
            .occurrences
            .iter()
            .find(|item| item.sequence > after && matcher.matches(item)))
    }

    /// Resolve an actual occurrence; an alias or command completion is not an anchor.
    pub fn anchor(&self, sequence: u64) -> Result<&Occurrence, Error> {
        self.healthy()?;
        self.occurrences
            .iter()
            .find(|item| item.sequence == sequence)
            .ok_or(Error::UnknownAnchor(sequence))
    }

    /// Assert strict observation order, including distinct occurrences in one clock tick.
    pub fn ordered(&self, before: u64, after: u64) -> Result<(), Error> {
        self.anchor(before)?;
        self.anchor(after)?;
        if before >= after {
            return Err(Error::Order { before, after });
        }
        Ok(())
    }

    fn end(&self, anchor: u64, duration_ms: u64) -> Result<u64, Error> {
        let start = self.anchor(anchor)?;
        if duration_ms == 0 {
            return Err(Error::InvalidWindow);
        }
        start
            .at_ms
            .checked_add(duration_ms)
            .ok_or(Error::InvalidWindow)
    }

    fn complete(&self, required_ms: u64) -> Result<(), Error> {
        self.healthy()?;
        if self.complete_before_ms < required_ms {
            return Err(Error::Unfinished {
                required_ms,
                observed_ms: self.complete_before_ms,
            });
        }
        Ok(())
    }

    /// Require no matching occurrence after the anchor and before its exclusive end.
    /// An offer exactly at wrap-up expiry is allowed. An early counterexample refutes
    /// immediately; success still requires the whole interval to be observed.
    pub fn absent(&self, matcher: &Matcher, anchor: u64, duration_ms: u64) -> Result<(), Error> {
        let end = self.end(anchor, duration_ms)?;
        if let Some(item) = self
            .occurrences
            .iter()
            .find(|item| item.sequence > anchor && item.at_ms < end && matcher.matches(item))
        {
            return Err(Error::Counterexample(item.clone()));
        }
        self.complete(end)
    }

    /// Require the baseline and every scoped occurrence before the exclusive end to
    /// satisfy the required payload. `scope` selects the account/queue stream; `required`
    /// adds the claimed metric values. A transient zero cannot hide between endpoints.
    pub fn stable(
        &self,
        scope: &Matcher,
        required: &BTreeMap<String, Node>,
        anchor: u64,
        duration_ms: u64,
    ) -> Result<(), Error> {
        let baseline = self.anchor(anchor)?;
        let end = self.end(anchor, duration_ms)?;
        if required.is_empty() {
            return Err(Error::EmptyClaim);
        }
        if required.keys().any(|field| {
            scope
                .payload
                .keys()
                .any(|scoped| overlapping_paths(field, scoped))
        }) {
            return Err(Error::FilteredClaim);
        }
        if !scope.matches(baseline) {
            return Err(Error::Counterexample(baseline.clone()));
        }
        for item in self
            .occurrences
            .iter()
            .filter(|item| item.sequence >= anchor && item.at_ms < end && scope.matches(item))
        {
            if required
                .iter()
                .any(|(field, value)| path_value(&item.payload, field) != Some(value))
            {
                return Err(Error::Counterexample(item.clone()));
            }
        }
        self.complete(end)
    }
}
